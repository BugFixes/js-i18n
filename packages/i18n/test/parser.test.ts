import parse from '../src/utils/parser'

describe('Parser', function () {
  test('correctly parses a plain translation body', function () {
    const output = parse('The quick brown fox jumps over the lazy dog')

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated identifier', function () {
    const output = parse('The quick ${animalColor} fox jumps over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated string', function () {
    const output = parse('The quick ${\'brown\'} fox jumps over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated string template', function () {
    const output = parse('The ${`quick ${animalColor} ${animalSpecies}`} jumps over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated number', function () {
    const output = parse('The quick brown fox jumps ${2.5} meters over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated boolean', function () {
    const output = parse('It is ${false} that the quick brown fox jumps over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated null value', function () {
    const output = parse('This is a ${null} interpolation') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated undefined value', function () {
    const output = parse('This is an ${undefined} interpolation') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated function call', function () {
    const output = parse('The quick ${getFoxColor()} fox jumps over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated function call with arguments', function () {
    const output = parse('The quick ${getAnimal(\'fox\', \'brown\')} jumps over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated function call with arguments containing shorthand `undefined` values', function () {
    const output = parse('The quick ${getAnimal(,,)} jumps over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated function call with an array as argument', function () {
    const output = parse('You have ${getTotal([10, 14, 3])} total messages') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated function call with an array containing shorthand `undefined` values as argument', function () {
    const output = parse('You have ${getTotal([,,])} total messages') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('correctly parses a translation body with an interpolated function call with an object as argument', function () {
    const output = parse('The quick ${getAnimal({ color: \'brown\', species: \'fox\' })} jumps over the lazy dog') // eslint-disable-line no-template-curly-in-string

    expect(output).toMatchSnapshot()
  })

  test('throws an error when the syntax is invalid', function () {
    // Expression start without matching closing curly braces
    expect(function () {
      parse('You have ${ total messages')
    }).toThrow('Invalid translation syntax')

    // Object without matching closing curly braces
    expect(function () {
      parse('You have ${{} total messages') // eslint-disable-line no-template-curly-in-string
    }).toThrow('Invalid translation syntax')

    // Object with shorthand property declaration
    expect(function () {
      parse('You have ${{h}} total messages') // eslint-disable-line no-template-curly-in-string
    }).toThrow('Invalid translation syntax')

    // Template literal expression start without matching closing curly braces
    expect(function () {
      parse('You have ${`f${bar`} total messages') // eslint-disable-line no-template-curly-in-string
    }).toThrow('Invalid translation syntax')

    // Template literal closing curly braces without matching opening curly braces
    expect(function () {
      parse('You have ${`f${}}bar`} total messages') // eslint-disable-line no-template-curly-in-string
    }).toThrow('Invalid translation syntax')
  })
})
