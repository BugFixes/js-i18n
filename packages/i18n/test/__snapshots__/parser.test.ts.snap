// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser correctly parses a plain translation body 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "The quick brown fox jumps over the lazy dog",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated boolean 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "It is ",
  },
  Object {
    "type": "Literal",
    "value": false,
  },
  Object {
    "type": "Literal",
    "value": " that the quick brown fox jumps over the lazy dog",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated function call 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "The quick ",
  },
  Object {
    "arguments": Array [],
    "callee": "getFoxColor",
    "type": "CallExpression",
  },
  Object {
    "type": "Literal",
    "value": " fox jumps over the lazy dog",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated function call with an array as argument 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "You have ",
  },
  Object {
    "arguments": Array [
      Object {
        "elements": Array [
          Object {
            "type": "Literal",
            "value": 10,
          },
          Object {
            "type": "Literal",
            "value": 14,
          },
          Object {
            "type": "Literal",
            "value": 3,
          },
        ],
        "type": "ArrayExpression",
      },
    ],
    "callee": "getTotal",
    "type": "CallExpression",
  },
  Object {
    "type": "Literal",
    "value": " total messages",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated function call with an array containing shorthand \`undefined\` values as argument 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "You have ",
  },
  Object {
    "arguments": Array [
      Object {
        "elements": Array [
          Object {
            "type": "Literal",
            "value": undefined,
          },
          Object {
            "type": "Literal",
            "value": undefined,
          },
        ],
        "type": "ArrayExpression",
      },
    ],
    "callee": "getTotal",
    "type": "CallExpression",
  },
  Object {
    "type": "Literal",
    "value": " total messages",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated function call with an object as argument 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "The quick ",
  },
  Object {
    "arguments": Array [
      Object {
        "properties": Object {
          "color": Object {
            "type": "Literal",
            "value": "brown",
          },
          "species": Object {
            "type": "Literal",
            "value": "fox",
          },
        },
        "type": "ObjectExpression",
      },
    ],
    "callee": "getAnimal",
    "type": "CallExpression",
  },
  Object {
    "type": "Literal",
    "value": " jumps over the lazy dog",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated function call with arguments 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "The quick ",
  },
  Object {
    "arguments": Array [
      Object {
        "type": "Literal",
        "value": "fox",
      },
      Object {
        "type": "Literal",
        "value": "brown",
      },
    ],
    "callee": "getAnimal",
    "type": "CallExpression",
  },
  Object {
    "type": "Literal",
    "value": " jumps over the lazy dog",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated function call with arguments containing shorthand \`undefined\` values 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "The quick ",
  },
  Object {
    "arguments": Array [
      Object {
        "type": "Literal",
        "value": undefined,
      },
      Object {
        "type": "Literal",
        "value": undefined,
      },
      Object {
        "type": "Literal",
        "value": undefined,
      },
    ],
    "callee": "getAnimal",
    "type": "CallExpression",
  },
  Object {
    "type": "Literal",
    "value": " jumps over the lazy dog",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated identifier 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "The quick ",
  },
  Object {
    "name": "animalColor",
    "type": "Identifier",
  },
  Object {
    "type": "Literal",
    "value": " fox jumps over the lazy dog",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated null value 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "This is a ",
  },
  Object {
    "type": "Literal",
    "value": null,
  },
  Object {
    "type": "Literal",
    "value": " interpolation",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated number 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "The quick brown fox jumps ",
  },
  Object {
    "type": "Literal",
    "value": 2.5,
  },
  Object {
    "type": "Literal",
    "value": " meters over the lazy dog",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated string 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "The quick ",
  },
  Object {
    "type": "Literal",
    "value": "brown",
  },
  Object {
    "type": "Literal",
    "value": " fox jumps over the lazy dog",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated string template 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "The ",
  },
  Object {
    "parts": Array [
      Object {
        "type": "Literal",
        "value": "quick ",
      },
      Object {
        "name": "animalColor",
        "type": "Identifier",
      },
      Object {
        "type": "Literal",
        "value": " ",
      },
      Object {
        "name": "animalSpecies",
        "type": "Identifier",
      },
    ],
    "type": "TemplateLiteral",
  },
  Object {
    "type": "Literal",
    "value": " jumps over the lazy dog",
  },
]
`;

exports[`Parser correctly parses a translation body with an interpolated undefined value 1`] = `
Array [
  Object {
    "type": "Literal",
    "value": "This is an ",
  },
  Object {
    "type": "Literal",
    "value": undefined,
  },
  Object {
    "type": "Literal",
    "value": " interpolation",
  },
]
`;
